lib/ansible/modules/cloud/vmware/vmware_host_config_manager.py

```diff
@@ -98,18 +99,70 @@ def __init__(self, module):
-                host_facts[option.key] = option.value
+                host_facts[option.key] = dict(value=option.value)
+
+            for s_option in option_manager.supportedOption:
+                host_facts[s_option.key].update(
+                    option_type=s_option.optionType,
+                )
+
             change_option_list = []
             for option_key, option_value in self.options.items():
-                if option_key in host_facts and option_value != host_facts[option_key]:
-                    change_option_list.append(vim.option.OptionValue(key=option_key, value=option_value))
-                    changed = True
+                if option_key in host_facts:
+                    # Make sure option_type is defined some values do not have
+                    # it defined and appear to be read only.
+                    if 'option_type' in host_facts[option_key]:
+                        # We handle all supported types here so we can give meaningful errors.
+                        option_type = host_facts[option_key]['option_type']
+                        if self.is_boolean(option_value) and isinstance(option_type, vim.option.BoolOption):
+                            option_value = self.is_truthy(option_value)
+                        elif (isinstance(option_value, integer_types) or self.is_integer(option_value))\
+                                and isinstance(option_type, vim.option.IntOption):
+                            option_value = VmomiSupport.vmodlTypes['int'](option_value)
+                        elif (isinstance(option_value, integer_types) or self.is_integer(option_value, 'long'))\
+                                and isinstance(option_type, vim.option.LongOption):
+                            option_value = VmomiSupport.vmodlTypes['long'](option_value)
+                        elif isinstance(option_value, float) and isinstance(option_type, vim.option.FloatOption):
+                            pass
+                        elif isinstance(option_value, string_types) and isinstance(option_type, (vim.option.StringOption, vim.option.ChoiceOption)):
+                            pass
+                        else:
+                            self.module.fail_json(msg="Provided value is of type %s."
+                                                      " Option %s expects: %s" % (type(option_value), option_key, type(option_type)))
+                    else:
+                        self.module.fail_json(msg="Cannot change read only option %s to %s." % (option_key, option_value))
+
+                    if option_value != host_facts[option_key]['value']:
+                        change_option_list.append(vim.option.OptionValue(key=option_key, value=option_value))
+                        changed = True
+                else:  # Don't silently drop unknown options. This prevents typos from falling through the cracks.
+                    self.module.fail_json(msg="Unknown option %s" % option_key)
```

Comment:
Since we are handling unknown names here we should probably add "This is probably a bug." to the  "except vim.fault.InvalidName as e:" error below.

Reply:
OK.